pkg/autoscaler/stats_scraper.go

```diff
@@ -0,0 +1,142 @@
+limitations under the License.
+*/
+
+package autoscaler
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/knative/serving/pkg/apis/serving"
+	"github.com/knative/serving/pkg/reconciler"
+	"github.com/prometheus/common/expfmt"
+	"go.uber.org/zap"
+)
+
+// StatsScraper defines the interface for collecting Revision metrics
+type StatsScraper interface {
+	// Scrape scrapes the Revision queue metric endpoint then sends it as a
+	// StatMessage to the given channel.
+	Scrape(statsCh chan<- *StatMessage)
+}
+
+// cacheDisabledClient is a http client with cache disabled. It is shared by
+// every goruntime for a revision scraper.
+var cacheDisabledClient = &http.Client{
+	Transport: &http.Transport{
+		// Do not use the cached connection
+		DisableKeepAlives: true,
+	},
+	Timeout: 3 * time.Second,
+}
+
+// ServiceScraper scrapes Revision metrics via a K8S service by sampling. Which
+// pod to be picked up to serve the request is decided by K8S.
+type ServiceScraper struct {
+	httpClient *http.Client
+	url        string
+	metricKey  string
+	logger     *zap.SugaredLogger
+}
+
+// NewServiceScraper creates a new StatsScraper for the Revision which
+// the given Metric is responsible for.
+func NewServiceScraper(metric *Metric, logger *zap.SugaredLogger) (*ServiceScraper, error) {
+	return newServiceScraperWithClient(metric, logger, cacheDisabledClient)
+}
+
+func newServiceScraperWithClient(metric *Metric, logger *zap.SugaredLogger, httpClient *http.Client) (*ServiceScraper, error) {
+	revName := metric.Labels[serving.RevisionLabelKey]
+	if revName == "" {
+		return nil, fmt.Errorf("no Revision label found for Metric %s", metric.Name)
+	}
+
+	serviceName := reconciler.GetServingK8SServiceNameForObj(revName)
+	return &ServiceScraper{
+		httpClient: httpClient,
+		url:        fmt.Sprintf("http://%s.%s:9090/metrics", serviceName, metric.Namespace),
+		metricKey:  NewMetricKey(metric.Namespace, metric.Name),
+		logger:     logger,
+	}, nil
+}
+
+// Scrape call the destination service then send it
+// to the given stats chanel
+func (s *ServiceScraper) Scrape(statsCh chan<- *StatMessage) {
+	stat, err := s.scrapeViaURL()
+	if err != nil {
+		s.logger.Errorw("Failed to get metrics", zap.Error(err))
+		return
+	}
+
+	s.sendStatMessage(*stat, statsCh)
+}
+
+func (s *ServiceScraper) scrapeViaURL() (*Stat, error) {
+	req, err := http.NewRequest("GET", s.url, nil)
+	resp, err := s.httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return nil, fmt.Errorf("GET request for URL %q returned HTTP status %v", s.url, resp.StatusCode)
+	}
+
+	return extractData(resp)
+}
+
+func extractData(resp *http.Response) (*Stat, error) {
+	stat := Stat{}
+	var parser expfmt.TextParser
+	metricFamilies, err := parser.TextToMetricFamilies(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("Reading text format failed: %v", err)
+	}
+
+	now := time.Now()
```

Comment:
this variable is unsused except for the next line, so let's just remove it.

Reply:
I could't get a point of `time.Now()`. It has to be set to a variable first.